xds/server_security_ext_test.go

```
@@ -214,25 +215,38 @@ func (s) TestServer_Security_NoCertificateProvidersInBootstrap_Failure(t *testin
 		t.Fatal("Timeout when waiting for an NACK from the xDS client for the LDS response")
 	}
 
-	// Wait a short duration and ensure that the server does not enter "serving"
-	// mode.
-	sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
-	defer sCancel()
-	select {
-	case <-sCtx.Done():
-	case <-modeChangeHandler.modeCh:
-		t.Fatal("Server started serving RPCs before the route config was received")
-	}
-
-	// Create a client that uses insecure creds and verify that RPCs don't
-	// succeed.
-	cc, err := grpc.NewClient(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
-	if err != nil {
-		t.Fatalf("Failed to dial local test server: %v", err)
+	// Since the listener resource from the management server is invalid and is
+	// not cached, it should cause the server to go to NOT_SERVING and the
+	// error message should contain the xDS node ID.
+	//
+	// Even though the server is currently NOT_SERVING, in the case (where we
+	// are NOT_SERVING and the new mode is also NOT_SERVING), the update is not
+	// suppressed as:
+	//   1. the error may have change
+	//   2. it provides a timestamp of the last backoff attempt
+	// The loop ensures that we drain all the mode change notifications but
+	// at the same time verify that the server is NOT_SERVING and the error
+	// message contains the xDS node ID.
+	exit := false
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fatalf("Timed out waiting for server to go NOT_SERVING")
+		case gotMode := <-modeChangeHandler.modeCh:
+			if gotMode != connectivity.ServingModeNotServing {
+				t.Fatalf("Mode changed to %v, want %v", gotMode, connectivity.ServingModeNotServing)
+			}
+			gotErr := <-modeChangeHandler.errCh
+			if gotErr == nil || !strings.Contains(gotErr.Error(), nodeID) {
+				t.Fatalf("Unexpected error: %v, want xDS Node id: %s", gotErr, nodeID)
+			}
+		default:
+			exit = true
```

This is racy. You cout hit the "default" case for the `select` the first time around when your server has still not switched the mode to "not serving" and the test will pass incorrectly.

---

Update the test handler to suppress error updates for no_serving -> no_serving cases.